src/NServiceBus.AcceptanceTests/Performance/When_message_is_faulted.cs

```diff
@@ -0,0 +1,138 @@
+            Assert.That(processingStarted, Is.LessThanOrEqualTo(processingEnded));
+            Assert.IsTrue(context.IsMessageHandledByTheFaultEndpoint);
+        }
+
+        public class Context : ScenarioContext
+        {
+            public bool IsMessageHandledByTheAuditEndpoint { get; set; }
+            public bool IsMessageHandledByTheFaultEndpoint { get; set; }
+            public IReadOnlyDictionary<string, string> Headers { get; set; }
+            public IReadOnlyDictionary<string, string> FaultHeaders { get; set; }
+        }
+
+        public class EndpointWithAuditOn : EndpointConfigurationBuilder
+        {
+            public EndpointWithAuditOn()
+            {
+                EndpointSetup<DefaultServer>((c, r) =>
+                {
+                    c.SendFailedMessagesTo("errorQueueForAcceptanceTest");
+                }).AuditTo<EndpointThatHandlesAuditMessages>();
+            }
+
+            class MessageToBeAuditedHandler : IHandleMessages<MessageToBeAudited>
+            {
+                public Task Handle(MessageToBeAudited message, IMessageHandlerContext context)
+                {
+                    return context.SendLocal(new MessageThatFails());
+                }
+
+                public class MessageSentInsideHandlersHandler : IHandleMessages<MessageThatFails>
+                {
+                    public Context TestContext { get; set; }
+
+                    public Task Handle(MessageThatFails message, IMessageHandlerContext context)
+                    {
+                        throw new SimulatedException();
+                    }
+                }
+            }
+        }
+
+        public class EndpointThatHandlesAuditMessages : EndpointConfigurationBuilder
+        {
+            public EndpointThatHandlesAuditMessages()
+            {
+                EndpointSetup<DefaultServer>();
+            }
+
+            class AuditMessageHandler : IHandleMessages<MessageToBeAudited>
+            {
+                Context testContext;
+
+                public AuditMessageHandler(Context testContext)
+                {
+                    this.testContext = testContext;
+                }
+
+                public Task Handle(MessageToBeAudited message, IMessageHandlerContext context)
+                {
+                    testContext.Headers = context.MessageHeaders;
+                    testContext.IsMessageHandledByTheAuditEndpoint = true;
+                    return Task.FromResult(0);
+                }
+            }
+        }
+
+        class EndpointThatHandlesErrorMessages : EndpointConfigurationBuilder
+        {
+            public EndpointThatHandlesErrorMessages()
+            {
+                EndpointSetup<DefaultServer>()
+                    .CustomEndpointName("errorQueueForAcceptanceTest");
+            }
+
+            class ErrorMessageHandler : IHandleMessages<MessageThatFails>
+            {
+                Context testContext;
```

Comment:
This should be at the bottom of the class, try to do a full format in R#? (Ctrl+E+C)